pkg/controllers/provisioningv2/harvestercleanup/controller.go

```diff
@@ -0,0 +1,159 @@
+
+	capiCluster, err := capr.GetCAPIClusterFromLabel(obj, h.capiClusters)
+	if err != nil && !apierrors.IsNotFound(err) {
+		logrus.Errorf("[%s] %s: error getting cluster (apiversion=cluster.x-k8s.io): %v", logMessageIdentifier, key, err)
+		return obj, err
+	}
+	if err != nil && apierrors.IsNotFound(err) {
+		logrus.Debugf("[%s] %s: there is no cluster (apiversion=cluster.x-k8s.io). Continue ...", logMessageIdentifier, key)
+		return obj, nil
+	}
+
+	// Skip if the CAPI cluster (cluster.x-k8s.io) is not being deleted.
+	if capiCluster.DeletionTimestamp.IsZero() {
+		return obj, nil
+	}
+
+	cluster, err := h.clusterCache.Get(objMeta.GetNamespace(), capiCluster.Name)
+	if err != nil {
+		logrus.Errorf("[%s] %s: error getting cluster (apiversion=provisioning.cattle.io): %v", logMessageIdentifier, key, err)
+		return obj, err
+	}
+
+	secret, err := machineprovision.GetCloudCredentialSecret(h.secretCache, "", cluster.Spec.CloudCredentialSecretName)
+	if err != nil {
+		logrus.Errorf("[%s] %s: error getting cloud credential secret: %v", logMessageIdentifier, key, err)
+		return obj, err
+	}
+
+	restConfig, err := clientcmd.RESTConfigFromKubeConfig(secret.Data["harvestercredentialConfig-kubeconfigContent"])
+	if err != nil {
+		logrus.Errorf("[%s] %s: error getting REST config from cloud credential secret: %v", logMessageIdentifier, key, err)
+		return obj, err
+	}
+
+	// We need to use the dynamic client here because we do not want to import
+	// the kubevirt API into Rancher.
+	dynamicClient, err := k8dynamic.NewForConfig(restConfig)
+	if err != nil {
+		logrus.Errorf("[%s] %s: error getting dynamic client: %v", logMessageIdentifier, key, err)
+		return obj, err
+	}
+
+	resourceGVR := schema.GroupVersionResource{
+		Group: "kubevirt.io", Version: "v1", Resource: "virtualmachines",
+	}
+	resourceName := objMeta.GetName()
+	resourceNamespace := objData.String("spec", "vmNamespace")
+
+	resource, err := dynamicClient.Resource(resourceGVR).Namespace(resourceNamespace).Get(context.TODO(), resourceName, metav1.GetOptions{})
+	if err != nil && !apierrors.IsNotFound(err) {
+		logrus.Errorf("[%s] %s: error getting VM (gvr=%s, namespace=%s, name=%s): %v",
+			logMessageIdentifier, key, resourceGVR.String(), resourceNamespace, resourceName, err)
+		return obj, err
+	}
+	if err != nil && apierrors.IsNotFound(err) {
+		logrus.Debugf("[%s] %s: there is no VM (gvr=%s, namespace=%s, name=%s). Continue ...",
+			logMessageIdentifier, key, resourceGVR.String(), resourceNamespace, resourceName)
+		return obj, nil
+	}
+
+	annotations := resource.GetAnnotations()
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+	if _, ok := annotations[removedAllPVCsAnnotationKey]; !ok {
+		annotations[removedAllPVCsAnnotationKey] = "true"
+		resource.SetAnnotations(annotations)
+
+		_, err = dynamicClient.Resource(resourceGVR).Namespace(resourceNamespace).Update(context.TODO(), resource, metav1.UpdateOptions{})
```

Comment:
This should take into account if the object itself is already being deleted, because IIRC you won't be able to update the annotation then. Might not be able to handle it here, but food for thought.